dtree.c

```diff
@@ -661,15 +743,53 @@ void handleTextInput(SDL_Event *event){
     }
 }
 
+void moveCursorLine(int relative_line){
+    char** lines = getLines(GRAPH.selected->text.buf, 1);
+
+    int len_so_far = 0;
+    for (int i = 0; lines[i]; i++) {
+            int col = CURSOR_POSITION - len_so_far;
+            // Move Up
+            if ( relative_line == -1 && i > 0 && CURSOR_POSITION >= len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])){
+                if ( col > strlen(lines[i-1]) ) {
+                    CURSOR_POSITION -= col + 1 + 1;
+                }
+                else {
+                    CURSOR_POSITION -= col + 1;                  // move to end of line - 1
+                    CURSOR_POSITION -= strlen(lines[i-1]) - col; //move to col'th idx of line i-1
+                }
+            }
+            // Move Down
+            else if ( relative_line == 1 && lines[i+1] && CURSOR_POSITION > len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])){
+                if ( col > strlen(lines[i+1]) ) {
+                    CURSOR_POSITION += strlen(lines[i]) - col + strlen(lines[i+1]);
+                }
+                else {
+                    CURSOR_POSITION += strlen(lines[i]) + 1 - col;                  // move to end of line - 1
+                    CURSOR_POSITION += col; //move to col'th idx of line i-1
+                }
+                break;
+            }
+        len_so_far += strlen(lines[i]) + 1;
+    }
+    for (int i = 0; lines[i]; i++)
+        free(lines[i]);
+    free(lines);
+
+}
+
 void doKeyDown(SDL_KeyboardEvent *event) {
-    if ( isEditMode(MODE) ){
-        switch(event->keysym.sym) {
-            case SDLK_BACKSPACE:
-                if ( CURRENT_BUFFER && CURRENT_BUFFER->len > 0)
-                    CURRENT_BUFFER->buf[--CURRENT_BUFFER->len] = '\0';
-        }
+    switch(event->keysym.sym) {
+        case SDLK_BACKSPACE: if ( isEditMode(MODE) ) deleteCharInBufferRelativeToCursor(0); return;
+        case SDLK_DELETE:   if ( isEditMode(MODE) ) deleteCharInBufferRelativeToCursor(1); return;
+        case SDLK_MINUS:     GRAPH_SCALE *= (9./10.); return;
+        case SDLK_EQUALS:    GRAPH_SCALE *= (10./9.); return;
```

Comment:
Perhaps zoom rate can be stored as a user customizable variable, with the inverse being inferred (e.g. `ZOOM_SPEED = 10./9.` and so for zooming out we would do `1./ZOOM_SPEED`. Multiplying the scale to increase/decrease seems a bit weird to me since it means zooming is exponential (e.g. the more zoomed in you are, the more you will zoom in). I was expecting something more like `GRAPH_SCALE += ZOOM_SPEED`.

Reply:
So the reason we multiply instead of add is that multiplying approaches zero at finer and finer intervals as the user zooms out, whereas if we subtracted something like `0.1` every time, if they were at `0.1` scale, they would just break everything if they tried to zoom out. Good point about using inverses though.